src/zope/app/form/testing.py

```diff
@@ -19,9 +19,44 @@
 __docformat__ = "reStructuredText"
 
 import os
-from zope.app.testing.functional import ZCMLLayer
+import zope.interface
+import zope.component
+from zope.publisher.browser import IBrowserRequest
+from zope.publisher.interfaces import IDefaultViewName
+from zope.publisher.interfaces.browser import IDefaultBrowserLayer
+from zope.traversing.interfaces import ITraversable
 
-AppFormLayer = ZCMLLayer(
-    os.path.join(os.path.split(__file__)[0], 'ftesting.zcml'),
-    __name__, 'AppFormLayer', allow_teardown=True)
+import zope.app.form
+import zope.app.wsgi.testlayer
 
+AppFormLayer = zope.app.wsgi.testlayer.BrowserLayer(zope.app.form)
+
+def browserView(for_, name, factory, layer=IDefaultBrowserLayer,
+                providing=zope.interface.Interface):
+    """Define a global browser view
+    """
+    provideAdapter(for_, providing, factory, name, (layer,))
+
+def browserViewProviding(for_, factory, providing, layer=IDefaultBrowserLayer):
+    """Define a view providing a particular interface."""
+    return browserView(for_, '', factory, layer, providing)
+
+
+stypes = list, tuple
+def provideAdapter(required, provided, factory, name='', with_=(), **kw):
+    if with_ is None and kw.has_key('with'): # pragma: no cover
```

Comment:
Replace `kw.has_key('with')` by `'with' in kw` to be Python 3 compatible?
Or remove it because it seems to be there for Python 2.4 compatibility (before `with` statements got introduced).